easybuild/tools/easystack.py

```diff
@@ -0,0 +1,186 @@
+        ec_names = []
+        for sw in self.software_list:
+            ec_to_append = '%s-%s-%s-%s.eb' % (str(sw.software), str(sw.version),
+                                               str(sw.toolchain_name), str(sw.toolchain_version))
+            if ec_to_append is None:
+                continue
+            else:
+                ec_names.append(ec_to_append)
+        return ec_names
+
+    # flags applicable to all sw (i.e. robot)
+    def get_general_options(self):
+        """Returns general options (flags applicable to all sw (i.e. --robot))"""
+        general_options = {}
+        # TODO add support for general_options
+        # general_options['robot'] = self.robot
+        # general_options['easybuild_version'] = self.easybuild_version
+        return general_options
+
+
+class SoftwareSpecs(object):
+    """Contains information about every SW that should be installed"""
+    def __init__(self, software, version, toolchain, toolchain_version, toolchain_name):
+        self.software = software
+        self.version = version
+        self.toolchain = toolchain
+        self.toolchain_version = toolchain_version
+        self.toolchain_name = toolchain_name
+        self.toolchain = toolchain
+        self.versionsuffix = None
+
+    # to be implemented
+    # def get_versionsuffix(self):
+    #     return self.versionsuffix or ''
+
+
+# implement this to your own needs - to create custom yaml/json/xml parser
+class GenericEasystackParser(object):
+    """Parent of all implemented parser classes"""
+    @ staticmethod
+    def parse(filepath):
+        """Parent of all implemented parser functions (i.e. JSON, XML...)"""
+        raise NotImplementedError
+
+
+class YamlEasystackParser(GenericEasystackParser):
+    """YAML file parser"""
+    @ staticmethod
+    def parse(filepath):
+        """Parses YAML file and assigns obtained values to SW config instances as well as general config instance"""
+        try:
+            with open(filepath, 'r') as f:
+                spec_dict = yaml.safe_load(f)
+
+                eb = Easystack()
+        except Exception:
+            raise EasyBuildError("Could not read provided easystack.")
+
+        sw_dict = spec_dict["software"]
+
+        # assign software-specific EB attributes
+        for software in sw_dict:
+            asterisk_err = "Easystack specifications of '%s' contain asterisk. " % (str(software))
+            asterisk_err += "Wildcard feature is not supported yet."
+            wrong_structure_err = "Easystack specifications of '%s' have wrong yaml structure." % (str(software))
+            try:
+                # iterates through toolchains to find out what sw version is needed
+                for yaml_tc in sw_dict[software]['toolchains']:
+                    # if version string containts asterisk or labels, raise error (not implemented yet)
+                    if '*' in str(sw_dict[software]['toolchains'][yaml_tc]['versions']):
```

Comment:
If you access the same part of the parsed file, use local variables.

Also, this check is in the wrong place, you're now just checking whether `'*'` is an entry in the list of versions, while you probably want to check whether `*` appears in *any* version?

```python
versions = toolchains[toolchain]['versions']
for version in versions:
    version = str(version)
    if '*' in version:

Reply:
i believe you are asking if this applies to i.e. `2.2.*`
it does, since the str() func takes the version (which might be of type dict or float at this point) and transfers it to string, which is iterable, therefore `if '*' in .....` is possible

Reply:
true, it returns error even if there is a correct version spec (without asterisk) alongside the incorrect one, but I suppose it is supposed to stop either way

Reply:
also, appears I did think about it. there is another condition that covers this - `if '*' in yaml_version:`
which is obsolete, when i think about it (just bcs of what i wrote in previous comments)